### ISSUE TITLE
How do I generate a quaternion from euler angles?

### ISSUE THREAD

**Issue by badasahog** [recp/cglm#306]
I want to get a quaternion based on roll, pitch, and yaw, but as far as I can tell there is no function for this.

The only functions I could find were based on a unit vector and an angle (glm_quatv, glm_quat)

glm provides this functionality in the form of the glm::quat(glm::vec3) constructor

**Comment by recp** [comment]
Hi @badasahog,

Sorry for the delay,

It could be nice to have this feature in **cglm** too. I had tried to update Euler with https://github.com/recp/cglm/pull/128 but couldn't finished that PR yet,

We can have this before that,

Any PR would be awesome otherwise I added to TODOs. 🙂

**Comment by Nonesence999** [comment]
Yes It would be amazing I also need a quat structure for my project.

**Comment by telephone001** [comment]
Sorry for answering very late, but for anyone else who needs help in the future, you could try making a function that rotates the quaternion by each axis. I guess you could also try and push it to the repository even though it might not be optimal

// example for rotating the quaternion by the y axis by an angle:

// y axis
vec3s axis = {0, 1, 0};

// creates a quaternion with the angle and axis
versors rotation = glms_quatv(angle, axis); 

// when multiplying this rotation with an angular position quaternion, it will apply the rotation to the angular position
angular_position = glms_quat_mul(rotation, angular_position);

**Comment by recp** [comment]
@telephone001 many thanks, hope we add this asap 🚀

**Comment by telephone001** [comment]
do you think glm_quate_xyz or glm_quat_euler_xyz would be a better name for this new function?

**Comment by recp** [comment]
To convert Euler to Quat, I would use `glm_euler_[x|y|z]_quat` which is similar pattern used in **cglm** e.g. glm_mat4_quat() convert mat4 to quat. 

```
glm_SRC_DST()
```

**Comment by telephone001** [comment]
Alright thanks for the response. I'll try and get this done soon

**Comment by recp** [comment]
@telephone001 thanks 🤗

**Comment by recp** [comment]
Thanks for they implementation a PR would be nice to see,

> it fails using meson webassembly for some reason.

https://github.com/recp/cglm/actions seems fine, may I ask what are the errors? Also can you pull latest commits and try again? there was an issue about https://github.com/recp/cglm/pull/360#issuecomment-1838122934 but fixed later.

**Comment by telephone001** [comment]
I created a pr and most tests failed. The error is just that all tests fail on certain builds.

**Comment by gottfriedleibniz** [comment]
The 'expected' quaternion is [overwriting](https://github.com/recp/cglm/blob/master/include/cglm/quat.h#L468) itself during multiplication.

Another thing to check is whether `glm_euler_*_quat` matches, or roughly matches, `glm_euler_by_order + glm_mat4_quat`. `GLM_EULER_XYZ` looks like `X*Y*Z` (local/intrinsic) instead of [`Z*Y*X`](https://github.com/recp/cglm/pull/369/files#diff-39ca64199733b629a85f4ba1acf35d353ee63ead5a535652b6807224ad328adfR130-R132) (global/extrinsic).

**Comment by telephone001** [comment]
Oh I see now. Thank you for the reply. I'll try and get this fixed up.
I'll also add glm_euler_by_order + glm_mat4_quat in tests
Also, I'm not really sure what you mean by local/intrinsic vs global/extrinsic

**Comment by recp** [comment]
> The 'expected' quaternion is [overwriting](https://github.com/recp/cglm/blob/5ae0cd4618b37049cf1ef78fc02bd65d0895b399/include/cglm/quat.h#L468-L471) itself during multiplication.

@gottfriedleibniz many thanks for the catch! 

I'll also check all places e.g. `dest[0]` to fix similar situations

> Also, I'm thinking of moving these functions to euler.h would that be better than quat.h?

Yes since SRC->DST is Euler -> Quat I think they must be in euler.h, In the future we can add reverse quat_to_euler in quat.h. 

> If we do settle on the xyz combo in the name, would it be allowed for me to make a macro function to create the xyz combination tests for me?

To simplify creating tests, any macro, utility functions can be used 👍
